File: Algorithm/Quick_sort.py
'''
Lomuto
'''

'''
Hoare
'''


def LomutoPartition(array: list, left: int, right: int) -> int:
    # right = len - 1
    pivot = array[right]
    i = left - 1

    for j in range(left, right):
        if array[j] <= pivot:
            i += 1
            array[j], array[i] = array[i], array[j]

    i += 1
    array[i], array[right] = array[right], array[i]
    return i


def HoarePartition(array: list, left: int, right: int) -> int:
    # right = len - 1
    pivot = array[left]
    i = left
    j = right
    while True:
        while array[i] < pivot:
            i += 1
        while array[j] > pivot:
            j -= 1
        if i < j:
            array[i], array[j] = array[j], array[i]
        else:
            return j


def QuickSort(array: list, left: int, right):
    # right = len - 1
    if left >= right:
        return
    # pivot = LomutoPartition(array, left, right)
    pivot = HoarePartition(array, left, right)

    # Recursion
    QuickSort(array, left, pivot - 1)
    QuickSort(array, pivot + 1, right)


data = [-2, 45, 0, 11, -9]
QuickSort(data, 0, len(data) - 1)
print(data)

File: Algorithm/QuickSort.cpp
#include <iostream>
using namespace std;

void swap(int a, int b) {
  int temp = b;
  b = a;
  a = temp;
}

// Partition function return the position of the pivot after partition process
int HoarePartition(int arr[], int left, int right) {
    bool found = false;
    int pivot = arr[left];
    int i = left; int j = right + 1;
    while (!found) {
        do {
            i++;
        } while (arr[i] < pivot);
        do {
            j--;
        } while (arr[j] < pivot);
        if (i >= j) found = true;
        swap(arr[i], arr[j]);
    }
    return -1;
}

int LomutoPartition(int arr[], int left, int right) {
    int i = left - 1;
    int pivot = arr[right - 1];
    for (int j = left; i < right; j++) {
      if (arr[j] < pivot) {
        i ++;
        swap(arr[i], arr[j]);
      }
    }
    swap(arr[i + 1], pivot);
    return i + 1;
}

void QuickSort(int arr[], int low, int high) {
  int pi;
  if (low < high) {
    //pi = LomutoPartition(arr, low, high);
    pi = HoarePartition(arr, low, high);
    QuickSort(arr, low, pi - 1);
    QuickSort(arr, pi, high);
  }
}

int main() {
  int arr[] = {1,2,10,5,6,3,7,8,4};
  QuickSort(arr, 0, 8);
  return 0;
}
